File: scripts/rebuild_regions.mjs
#!/usr/bin/env node
/**
 * rebuild_regions.mjs
 *
 * Rebuilds the regions table from data/regions_rebuild.json.
 * Two modes:
 *   --dry-run   Validate JSON, check countries exist in DB, report what would be inserted (default)
 *   --insert    Actually insert new regions into the DB (does NOT delete old regions)
 *
 * The appellation remapping and old region cleanup are deferred to a future session.
 *
 * Usage:
 *   node scripts/rebuild_regions.mjs              # Dry run
 *   node scripts/rebuild_regions.mjs --insert     # Insert new regions
 */

import { readFileSync } from 'fs';
import { createClient } from '@supabase/supabase-js';

// ── Load .env ───────────────────────────────────────────────
const envPath = new URL('../.env', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const envContent = readFileSync(envPath, 'utf-8');
for (const line of envContent.split('\n')) {
  const trimmed = line.replace(/\r/g, '').trim();
  if (!trimmed || trimmed.startsWith('#')) continue;
  const eqIdx = trimmed.indexOf('=');
  if (eqIdx === -1) continue;
  const key = trimmed.slice(0, eqIdx);
  const val = trimmed.slice(eqIdx + 1);
  if (!process.env[key]) process.env[key] = val;
}

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_ANON_KEY;
const sb = createClient(SUPABASE_URL, SUPABASE_KEY);

const INSERT_MODE = process.argv.includes('--insert');

// ── Step 1: Load and validate JSON ──────────────────────────
console.log('Step 1: Loading and validating data/regions_rebuild.json...\n');

const dataPath = new URL('../data/regions_rebuild.json', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const regions = JSON.parse(readFileSync(dataPath, 'utf8'));

const catchAll = regions.filter(r => r.is_catch_all);
const l1 = regions.filter(r => !r.is_catch_all && !r.parent);
const l2 = regions.filter(r => !r.is_catch_all && r.parent);

console.log(`  Total entries: ${regions.length}`);
console.log(`  Catch-all: ${catchAll.length}`);
console.log(`  L1: ${l1.length}`);
console.log(`  L2: ${l2.length}`);

// Check duplicate slugs
const slugs = regions.map(r => r.slug);
const seen = new Set();
const dupes = [];
for (const s of slugs) {
  if (seen.has(s)) dupes.push(s);
  seen.add(s);
}
if (dupes.length > 0) {
  console.error(`\n  ERROR: Duplicate slugs found: ${dupes.join(', ')}`);
  process.exit(1);
}
console.log('  Duplicate slugs: none ✓');

// Check L2 parents reference valid L1 slugs
const l1Slugs = new Set(l1.map(r => r.slug));
const badParents = l2.filter(r => !l1Slugs.has(r.parent));
if (badParents.length > 0) {
  console.error(`\n  ERROR: L2 regions with invalid parents:`);
  for (const r of badParents) console.error(`    ${r.slug} -> ${r.parent}`);
  process.exit(1);
}
console.log('  L2 parent references: all valid ✓');

// Check each country has exactly one catch-all
const countriesInData = new Set(regions.map(r => r.country));
const catchAllByCountry = {};
for (const r of catchAll) {
  catchAllByCountry[r.country] = (catchAllByCountry[r.country] || 0) + 1;
}
for (const c of countriesInData) {
  if (!catchAllByCountry[c]) {
    console.error(`\n  ERROR: Country "${c}" has no catch-all region`);
    process.exit(1);
  }
  if (catchAllByCountry[c] > 1) {
    console.error(`\n  ERROR: Country "${c}" has ${catchAllByCountry[c]} catch-all regions`);
    process.exit(1);
  }
}
console.log('  Catch-all per country: exactly one each ✓');

// ── Step 2: Check countries exist in DB ─────────────────────
console.log('\nStep 2: Verifying countries exist in database...\n');

// Fetch all countries from DB
const allCountries = [];
let from = 0;
const PAGE = 1000;
while (true) {
  const { data, error } = await sb.from('countries').select('id, name, slug').range(from, from + PAGE - 1);
  if (error) { console.error('  DB error:', error.message); process.exit(1); }
  allCountries.push(...data);
  if (data.length < PAGE) break;
  from += PAGE;
}

const countryByName = new Map(allCountries.map(c => [c.name, c]));
const missing = [];
for (const c of countriesInData) {
  if (!countryByName.has(c)) missing.push(c);
}
if (missing.length > 0) {
  console.error(`  ERROR: Countries not found in DB: ${missing.join(', ')}`);
  process.exit(1);
}
console.log(`  All ${countriesInData.size} countries found in DB ✓`);

// ── Step 3: Check for existing regions with same slugs ──────
console.log('\nStep 3: Checking for slug conflicts with existing regions...\n');

const existingRegions = [];
from = 0;
while (true) {
  const { data, error } = await sb.from('regions').select('id, name, slug, country_id, is_catch_all, parent_id').range(from, from + PAGE - 1);
  if (error) { console.error('  DB error:', error.message); process.exit(1); }
  existingRegions.push(...data);
  if (data.length < PAGE) break;
  from += PAGE;
}
console.log(`  Existing regions in DB: ${existingRegions.length}`);

const existingSlugSet = new Set(existingRegions.map(r => r.slug));
const conflicts = regions.filter(r => existingSlugSet.has(r.slug));
if (conflicts.length > 0) {
  const catchAllConflicts = conflicts.filter(r => r.is_catch_all);
  const nonCatchAllConflicts = conflicts.filter(r => !r.is_catch_all);
  console.log(`  Slug conflicts: ${conflicts.length} (${catchAllConflicts.length} catch-all, ${nonCatchAllConflicts.length} non-catch-all)`);
  if (nonCatchAllConflicts.length > 0) {
    console.log('  Non-catch-all conflicts:');
    for (const r of nonCatchAllConflicts.slice(0, 10)) {
      console.log(`    ${r.slug} (${r.country}/${r.name})`);
    }
    if (nonCatchAllConflicts.length > 10) console.log(`    ... and ${nonCatchAllConflicts.length - 10} more`);
  }
} else {
  console.log('  No slug conflicts ✓');
}

// ── Summary ─────────────────────────────────────────────────
console.log('\n' + '═'.repeat(60));
console.log('SUMMARY');
console.log('═'.repeat(60));
console.log(`  Would insert: ${regions.length - conflicts.length} new regions`);
console.log(`  Would skip (slug exists): ${conflicts.length}`);
console.log(`  Countries with new regions: ${[...countriesInData].filter(c => {
  return regions.filter(r => r.country === c && !r.is_catch_all && !existingSlugSet.has(r.slug)).length > 0;
}).length}`);

// Per-country breakdown
console.log('\n  Per-country breakdown:');
for (const c of [...countriesInData].sort()) {
  const regs = regions.filter(r => r.country === c && !r.is_catch_all);
  if (regs.length === 0) continue;
  const newRegs = regs.filter(r => !existingSlugSet.has(r.slug));
  const existingRegs = regs.filter(r => existingSlugSet.has(r.slug));
  console.log(`    ${c}: ${regs.length} regions (${newRegs.length} new, ${existingRegs.length} existing)`);
}

if (!INSERT_MODE) {
  console.log('\n  Mode: DRY RUN — no changes made');
  console.log('  Run with --insert to insert new regions into DB');
  process.exit(0);
}

// ── Step 4: Insert new regions ──────────────────────────────
console.log('\n' + '═'.repeat(60));
console.log('INSERTING NEW REGIONS');
console.log('═'.repeat(60));

// Build country_id lookup
const countryIdByName = new Map(allCountries.map(c => [c.name, c.id]));

// Insert catch-all regions first (skip existing)
let inserted = 0;
let skipped = 0;

console.log('\n  Inserting catch-all regions...');
for (const r of catchAll) {
  if (existingSlugSet.has(r.slug)) {
    skipped++;
    continue;
  }
  const { error } = await sb.from('regions').insert({
    name: r.name,
    slug: r.slug,
    country_id: countryIdByName.get(r.country),
    is_catch_all: true,
    parent_id: null,
  });
  if (error) {
    console.error(`    ERROR inserting catch-all ${r.slug}: ${error.message}`);
    continue;
  }
  inserted++;
}
console.log(`    Inserted: ${inserted}, Skipped (existing): ${skipped}`);

// Insert L1 regions
console.log('\n  Inserting L1 regions...');
let l1Inserted = 0;
let l1Skipped = 0;
const newL1IdBySlug = new Map();

for (const r of l1) {
  if (existingSlugSet.has(r.slug)) {
    // Fetch existing ID for L2 parent resolution
    const existing = existingRegions.find(e => e.slug === r.slug);
    if (existing) newL1IdBySlug.set(r.slug, existing.id);
    l1Skipped++;
    continue;
  }
  const { data, error } = await sb.from('regions').insert({
    name: r.name,
    slug: r.slug,
    country_id: countryIdByName.get(r.country),
    is_catch_all: false,
    parent_id: null,
  }).select('id');
  if (error) {
    console.error(`    ERROR inserting L1 ${r.slug}: ${error.message}`);
    continue;
  }
  newL1IdBySlug.set(r.slug, data[0].id);
  l1Inserted++;
}
console.log(`    Inserted: ${l1Inserted}, Skipped (existing): ${l1Skipped}`);

// Insert L2 regions
console.log('\n  Inserting L2 regions...');
let l2Inserted = 0;
let l2Skipped = 0;
let l2Errors = 0;

for (const r of l2) {
  if (existingSlugSet.has(r.slug)) {
    l2Skipped++;
    continue;
  }
  const parentId = newL1IdBySlug.get(r.parent);
  if (!parentId) {
    console.error(`    ERROR: No parent ID found for ${r.slug} -> ${r.parent}`);
    l2Errors++;
    continue;
  }
  const { error } = await sb.from('regions').insert({
    name: r.name,
    slug: r.slug,
    country_id: countryIdByName.get(r.country),
    is_catch_all: false,
    parent_id: parentId,
  });
  if (error) {
    console.error(`    ERROR inserting L2 ${r.slug}: ${error.message}`);
    l2Errors++;
    continue;
  }
  l2Inserted++;
}
console.log(`    Inserted: ${l2Inserted}, Skipped (existing): ${l2Skipped}, Errors: ${l2Errors}`);

// Final summary
console.log('\n' + '═'.repeat(60));
console.log('DONE');
console.log('═'.repeat(60));
console.log(`  Total inserted: ${inserted + l1Inserted + l2Inserted}`);
console.log(`  Total skipped: ${skipped + l1Skipped + l2Skipped}`);
console.log(`  Total errors: ${l2Errors}`);

// Verify final count
const { count } = await sb.from('regions').select('*', { count: 'exact', head: true });
console.log(`  Total regions in DB now: ${count}`);

File: pipeline/geo/rebuild_regions.py
"""
Rebuilds the regions table from data/regions_rebuild.json.

Two modes:
  --dry-run   Validate JSON, check countries exist in DB, report what would be inserted (default)
  --insert    Actually insert new regions into the DB (does NOT delete old regions)

Usage:
  python -m pipeline.geo.rebuild_regions
  python -m pipeline.geo.rebuild_regions --insert
"""

import sys
import json
import argparse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from pipeline.lib.db import get_supabase
from pipeline.geo.helpers import fetch_all_paginated

PROJECT_ROOT = Path(__file__).resolve().parents[2]


def main():
    parser = argparse.ArgumentParser(description="Rebuild regions from JSON")
    parser.add_argument("--insert", action="store_true")
    args = parser.parse_args()

    insert_mode = args.insert
    sb = get_supabase()

    # Step 1: Load and validate JSON
    print("Step 1: Loading and validating data/regions_rebuild.json...\n")
    data_path = PROJECT_ROOT / "data" / "regions_rebuild.json"
    regions = json.loads(data_path.read_text(encoding="utf-8"))

    catch_all = [r for r in regions if r.get("is_catch_all")]
    l1 = [r for r in regions if not r.get("is_catch_all") and not r.get("parent")]
    l2 = [r for r in regions if not r.get("is_catch_all") and r.get("parent")]

    print(f"  Total entries: {len(regions)}")
    print(f"  Catch-all: {len(catch_all)}")
    print(f"  L1: {len(l1)}")
    print(f"  L2: {len(l2)}")

    # Check duplicate slugs
    seen_slugs = set()
    dupes = []
    for r in regions:
        if r["slug"] in seen_slugs:
            dupes.append(r["slug"])
        seen_slugs.add(r["slug"])
    if dupes:
        print(f"\n  ERROR: Duplicate slugs found: {', '.join(dupes)}")
        sys.exit(1)
    print("  Duplicate slugs: none")

    # Check L2 parents reference valid L1 slugs
    l1_slugs = {r["slug"] for r in l1}
    bad_parents = [r for r in l2 if r["parent"] not in l1_slugs]
    if bad_parents:
        print("\n  ERROR: L2 regions with invalid parents:")
        for r in bad_parents:
            print(f"    {r['slug']} -> {r['parent']}")
        sys.exit(1)
    print("  L2 parent references: all valid")

    # Check each country has exactly one catch-all
    countries_in_data = {r["country"] for r in regions}
    ca_by_country = {}
    for r in catch_all:
        ca_by_country[r["country"]] = ca_by_country.get(r["country"], 0) + 1
    for c in countries_in_data:
        if not ca_by_country.get(c):
            print(f"\n  ERROR: Country \"{c}\" has no catch-all region")
            sys.exit(1)
        if ca_by_country[c] > 1:
            print(f"\n  ERROR: Country \"{c}\" has {ca_by_country[c]} catch-all regions")
            sys.exit(1)
    print("  Catch-all per country: exactly one each")

    # Step 2: Check countries exist in DB
    print("\nStep 2: Verifying countries exist in database...\n")
    all_countries = fetch_all_paginated(sb, "countries", "id, name, slug")
    country_by_name = {c["name"]: c for c in all_countries}

    missing = [c for c in countries_in_data if c not in country_by_name]
    if missing:
        print(f"  ERROR: Countries not found in DB: {', '.join(missing)}")
        sys.exit(1)
    print(f"  All {len(countries_in_data)} countries found in DB")

    # Step 3: Check for existing regions with same slugs
    print("\nStep 3: Checking for slug conflicts with existing regions...\n")
    existing_regions = fetch_all_paginated(
        sb, "regions", "id, name, slug, country_id, is_catch_all, parent_id"
    )
    print(f"  Existing regions in DB: {len(existing_regions)}")

    existing_slug_set = {r["slug"] for r in existing_regions}
    conflicts = [r for r in regions if r["slug"] in existing_slug_set]
    if conflicts:
        ca_conflicts = [r for r in conflicts if r.get("is_catch_all")]
        non_ca_conflicts = [r for r in conflicts if not r.get("is_catch_all")]
        print(f"  Slug conflicts: {len(conflicts)} ({len(ca_conflicts)} catch-all, {len(non_ca_conflicts)} non-catch-all)")
        if non_ca_conflicts:
            print("  Non-catch-all conflicts:")
            for r in non_ca_conflicts[:10]:
                print(f"    {r['slug']} ({r['country']}/{r['name']})")
            if len(non_ca_conflicts) > 10:
                print(f"    ... and {len(non_ca_conflicts) - 10} more")
    else:
        print("  No slug conflicts")

    # Summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    print(f"  Would insert: {len(regions) - len(conflicts)} new regions")
    print(f"  Would skip (slug exists): {len(conflicts)}")

    countries_with_new = [
        c for c in countries_in_data
        if any(r["country"] == c and not r.get("is_catch_all") and r["slug"] not in existing_slug_set for r in regions)
    ]
    print(f"  Countries with new regions: {len(countries_with_new)}")

    print("\n  Per-country breakdown:")
    for c in sorted(countries_in_data):
        regs = [r for r in regions if r["country"] == c and not r.get("is_catch_all")]
        if not regs:
            continue
        new_regs = [r for r in regs if r["slug"] not in existing_slug_set]
        existing_regs = [r for r in regs if r["slug"] in existing_slug_set]
        print(f"    {c}: {len(regs)} regions ({len(new_regs)} new, {len(existing_regs)} existing)")

    if not insert_mode:
        print("\n  Mode: DRY RUN -- no changes made")
        print("  Run with --insert to insert new regions into DB")
        return

    # Step 4: Insert new regions
    print("\n" + "=" * 60)
    print("INSERTING NEW REGIONS")
    print("=" * 60)

    country_id_by_name = {c["name"]: c["id"] for c in all_countries}

    # Insert catch-all regions first
    inserted = 0
    skipped = 0
    print("\n  Inserting catch-all regions...")
    for r in catch_all:
        if r["slug"] in existing_slug_set:
            skipped += 1
            continue
        sb.table("regions").insert({
            "name": r["name"],
            "slug": r["slug"],
            "country_id": country_id_by_name[r["country"]],
            "is_catch_all": True,
            "parent_id": None,
        }).execute()
        inserted += 1
    print(f"    Inserted: {inserted}, Skipped (existing): {skipped}")

    # Insert L1 regions
    print("\n  Inserting L1 regions...")
    l1_inserted = 0
    l1_skipped = 0
    new_l1_id_by_slug = {}

    for r in l1:
        if r["slug"] in existing_slug_set:
            existing = next((e for e in existing_regions if e["slug"] == r["slug"]), None)
            if existing:
                new_l1_id_by_slug[r["slug"]] = existing["id"]
            l1_skipped += 1
            continue
        result = (
            sb.table("regions")
            .insert({
                "name": r["name"],
                "slug": r["slug"],
                "country_id": country_id_by_name[r["country"]],
                "is_catch_all": False,
                "parent_id": None,
            })
            .select("id")
            .execute()
        )
        new_l1_id_by_slug[r["slug"]] = result.data[0]["id"]
        l1_inserted += 1
    print(f"    Inserted: {l1_inserted}, Skipped (existing): {l1_skipped}")

    # Insert L2 regions
    print("\n  Inserting L2 regions...")
    l2_inserted = 0
    l2_skipped = 0
    l2_errors = 0

    for r in l2:
        if r["slug"] in existing_slug_set:
            l2_skipped += 1
            continue
        parent_id = new_l1_id_by_slug.get(r["parent"])
        if not parent_id:
            print(f"    ERROR: No parent ID found for {r['slug']} -> {r['parent']}")
            l2_errors += 1
            continue
        try:
            sb.table("regions").insert({
                "name": r["name"],
                "slug": r["slug"],
                "country_id": country_id_by_name[r["country"]],
                "is_catch_all": False,
                "parent_id": parent_id,
            }).execute()
            l2_inserted += 1
        except Exception as e:
            print(f"    ERROR inserting L2 {r['slug']}: {e}")
            l2_errors += 1
    print(f"    Inserted: {l2_inserted}, Skipped (existing): {l2_skipped}, Errors: {l2_errors}")

    # Final summary
    print("\n" + "=" * 60)
    print("DONE")
    print("=" * 60)
    print(f"  Total inserted: {inserted + l1_inserted + l2_inserted}")
    print(f"  Total skipped: {skipped + l1_skipped + l2_skipped}")
    print(f"  Total errors: {l2_errors}")

    count_result = sb.table("regions").select("*", count="exact", head=True).execute()
    print(f"  Total regions in DB now: {count_result.count}")


if __name__ == "__main__":
    main()
